### PR TITLE
Fix evaluation of non-uniform rational B-splines (NURBS)

### PR DESCRIPTION
**Pull request by saidctb**
- Improve `eval_field` to compute weighted fields
- Add tests for the evaluation of `NurbsMapping`

### CODE REVIEW

File: psydac/fem/basic.py [hunk 1]
```diff
@@ -69,7 +69,7 @@ def is_product( self ):
     # Abstract interface: evaluation methods
     #---------------------------------------
     @abstractmethod
-    def eval_field( self, field, *eta ):
+    def eval_field( self, field, *eta , weights=None):
         """
         Evaluate field at location(s) eta.
```
Comment (yguclu) [inline psydac/fem/basic.py:74]: Please update docstring. It should explain that `weights.space == field.space`

File: psydac/fem/basic.py [hunk 2]
```diff
@@ -89,7 +89,7 @@ def eval_field( self, field, *eta ):
         """
 
     @abstractmethod
-    def eval_field_gradient( self, field, *eta ):
+    def eval_field_gradient( self, field, *eta , weights=None):
         """
         Evaluate field gradient at location(s) eta.
```
Comment (yguclu) [inline psydac/fem/basic.py:94]: Same comment as above

File: psydac/fem/basic.py [hunk 3]
```diff
@@ -230,19 +230,19 @@ def __getitem__(self, key):
         return self._fields[key]
 
     # ...
-    def __call__( self, *eta ):
+    def __call__( self, *eta , weights=None):
         """Evaluate field at location identified by logical coordinates eta."""
-        return self._space.eval_field( self, *eta )
+        return self._space.eval_field( self, *eta , weights=weights)
```
Comment (yguclu) [inline psydac/fem/basic.py:235]: Same comment as above

File: psydac/fem/basic.py [hunk 4]
```diff
@@ -230,19 +230,19 @@ def __getitem__(self, key):
...
 
     # ...
-    def gradient( self, *eta ):
+    def gradient( self, *eta , weights=None):
         """Evaluate gradient of field at location identified by logical coordinates eta."""
-        return self._space.eval_field_gradient( self, *eta )
+        return self._space.eval_field_gradient( self, *eta , weights=weights)
```
Comment (yguclu) [inline psydac/fem/basic.py:240]: Same comment as above

File: psydac/fem/basic.py [hunk 5]
```diff
@@ -230,19 +230,19 @@ def __getitem__(self, key):
...
         
     # ...
-    def divergence(self, *eta):
+    def divergence(self, *eta, weights=None):
         """Evaluate divergence of vector field at location identified by logical coordinates eta."""
-        return self._space.eval_field_divergence(self, *eta)
+        return self._space.eval_field_divergence(self, *eta, weights=weights)
```
Comment (yguclu) [inline psydac/fem/basic.py:245]: Same comment as above

File: psydac/fem/splines.py [hunk 6]
```diff
@@ -239,10 +239,15 @@ def eval_field( self, field, *eta ):
         if self.basis == 'M':
             basis *= self._scaling_array[index]
 
-        return np.dot( field.coeffs[index], basis )
+        coeffs = field.coeffs[index].copy()
+
+        if weights:
+            coeffs *= wieghts[index]
```
Comment (yguclu) [inline psydac/fem/splines.py:245]: Typo: wieghts -> weights

I suspect that this code is never tested. Can you add a 1D unit test then?

**Review by yguclu — CHANGES_REQUESTED** [review]
Excellent fix! I have a couple of comments

File: psydac/fem/basic.py (hunk 1 above)
Comment (saidctb) [reply psydac/fem/basic.py:74]: okay

File: psydac/fem/basic.py (hunk 2 above)
Comment (saidctb) [reply psydac/fem/basic.py:97]: done

File: psydac/fem/splines.py (hunk 6 above)
Comment (saidctb) [reply psydac/fem/splines.py]: yes this part is never tested, how do we add a 1d test using `NurbsMapping`

File: psydac/fem/basic.py (hunk 3 above)
Comment (yguclu) [reply psydac/fem/basic.py:241]: @saidctb Still missing

File: psydac/fem/basic.py (hunk 4 above)
Comment (saidctb) [reply psydac/fem/basic.py]: done

File: psydac/fem/basic.py (hunk 5 above)
Comment (saidctb) [reply psydac/fem/basic.py]: this case is not available yet, so I added a `NotImplementedError`

File: psydac/fem/basic.py (hunk 3 above)
Comment (saidctb) [reply psydac/fem/basic.py]: done

File: psydac/fem/basic.py (hunk 4 above)
Comment (yguclu) [reply psydac/fem/basic.py]: This is the base class, it is not its job to know whether the method exists or not. `NotImplementedError` is already raised by the subclasses, I believe.

File: psydac/fem/splines.py (hunk 6 above)
Comment (yguclu) [reply psydac/fem/splines.py]: It should not be much different than a 2D test... You can create a 1D NURBS using Igakit

File: psydac/fem/basic.py (hunk 5 above)
Comment (yguclu) [reply psydac/fem/basic.py]: This is the base class, it is not its job to know whether the method exists or not. NotImplementedError is already raised by the subclasses, I believe.

File: psydac/fem/basic.py (hunk 4 above)
Comment (saidctb) [reply psydac/fem/basic.py]: ah okay

File: psydac/fem/splines.py (hunk 6 above)
Comment (saidctb) [reply psydac/fem/splines.py]: we always return a TensorFemSpace when we use the `HDF5` files, so we will need to create everything by hand

File: psydac/fem/basic.py (hunk 5 above)
Comment (saidctb) [reply psydac/fem/basic.py]: done

**Review by yguclu — APPROVED** [review]

File: psydac/fem/splines.py (hunk 6 above)
Comment (yguclu) [reply psydac/fem/splines.py]: You're right, it does not really matter then.

**Pull request merged**